design/sketch.webidl

```diff
@@ -0,0 +1,532 @@
+    WebGPUCommandBuffer finishEncoding();
+
+    // Commands allowed outside of "passes"
+    void copyBufferToBuffer(WebGPUBuffer src,
+                            u32 srcOffset,
+                            WebGPUBuffer dst,
+                            u32 dstOffset,
+                            u32 size);
+    // TODO figure out all the arguments required for these
+    void copyBufferToTexture();
+    void copyTextureToBuffer();
+    void copyTextureToTexture();
+    void blit();
+
+    // Allowed in both compute and render passes
+    void setPushConstants(WebGPUShaderStageFlags stage,
+                          u32 offset,
+                          u32 count,
+                          ArrayBuffer data);
+    void setBindGroup(u32 index, WebGPUBindGroup bindGroup);
+    void setPipeline((WebGPUComputePipeline or WebGPURenderPipeline) pipeline);
+
+    // Compute pass commands
+    void beginComputePass();
+    void endComputePass();
+
+    void dispatch(u32 x, u32 y, u32 z);
+
+    // Render pass commands
+    void beginRenderPass(WebGPURenderPassDescriptor descriptor);
+    void endRenderPass();
+
+    void setBlendColor(float r, float g, float b, float a);
+    void setIndexBuffer(WebGPUBuffer buffer, u32 offset);
+    void setVertexBuffers(u32 startSlot, sequence<WebGPUBuffer> buffers, sequence<u32> offsets);
+
+    void draw(u32 vertexCount, u32 instanceCount, u32 firstVertex, u32 firstInstance);
+    void drawIndexed(u32 indexCount, u32 instanceCount, u32 firstIndex, u32 firstInstance);
+
+    // TODO add missing commands
+};
+
+// ****************************************************************************
+// OTHER (Fence, Queue SwapChain, Device)
+// ****************************************************************************
+
+// Fence
+interface WebGPUFence {
+    u32 getValue();
+};
+
+// Queue
+interface WebGPUQueue {
+    void submit(sequence<WebGPUCommandBuffer> buffers);
+    void updateFence(WebGPUFence fence, u32 value);
+};
+
+// SwapChain / RenderingContext
+dictionary WebGPUSwapChainDescriptor {
+    WebGPUTextureUsageFlags usage;
+    WebGPUTextureFormatEnum format;
+    u32 width;
+    u32 height;
+};
+
+interface WebGPUSwapChain {
+    void configure(WebGPUSwapChainDescriptor descriptor);
+    WebGPUTexture getNextTexture();
+    void present();
+};
+
+interface WebGPURenderingContext : WebGPUSwapChain {
+};
+
+// Device
+interface WebGPUDevice {
+    WebGPUBuffer createBuffer(WebGPUBufferDescriptor descriptor);
+    WebGPUTexture createTexture(WebGPUTextureDescriptor descriptor);
+    WebGPUSampler createSampler(WebGPUSamplerDescriptor descriptor);
+
+    WebGPUBindGroupLayout createBindGroupLayout(WebGPUBindGroupLayoutDescriptor descriptor);
+    WebGPUPipelineLayout createPipelineLayout(WebGPUPipelineLayoutDescriptor descriptor);
+    WebGPUBindGroup createBindGroup(WebGPUBindGroupDescriptor descriptor);
+
+    WebGPUBlendState createBlendState(WebGPUBlendStateDescriptor descriptor);
```

Comment:
It's desirable to be able to set blend state on per-color-attachment basis.

Reply:
This is already the case with the following:

```
dictionary WebGPURenderPipelineDescriptor : WebGPUPipelineDescriptorBase{
    // ...
    sequence<WebGPUBlendState> blendState;
    // ...
};
```

That said we'll have to discuss if we when to support it be default given it is an optional Vulkan feature.